### PR TITLE
Add sequence support for identity on GPU

### PR DESCRIPTION
**Pull request by RandySheriffH**
1. Enhance infrastructure to support sequence copy from host to device;
2. Enhance Identity op GPU to support tensor sequence.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/tensor/identity_op.h [hunk 1]
```diff
@@ -15,34 +16,57 @@ class IdentityOp final : public CudaKernel {
   }
 
   Status ComputeInternal(OpKernelContext* context) const override {
-    const Tensor* X = context->Input<Tensor>(0);
-    if (X == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
-    const TensorShape& shape = X->Shape();
-    Tensor* Y = context->Output(0, shape);
-    auto X_type = X->DataType();
+    auto X_ml_type = context->InputType(0);
+    if (DataTypeImpl::GetType<Tensor>() == X_ml_type) {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/tensor/identity_op.h:20]: >DataTypeImpl::GetType<Tensor>() == X_ml_type [](http://example.com/codeflow?start=8&length=44)

This only covers a base class. use:
if(X_ml_type->IsTensor())

File: onnxruntime/core/providers/cuda/tensor/identity_op.h [hunk 2]
```diff
@@ -15,34 +16,57 @@ class IdentityOp final : public CudaKernel {
...
+          // output is tied with the type of the input in Opset 7 whereas
+          // the type of 'mask' in Opset 10 is 'bool' always
+          // so we have a common solution
+          void* mask_data = mask->MutableDataRaw();
+          // In 'test'/'inference' mode, there are no input values dropped out
+          // so fill the buffer with 0/false
+          CUDA_RETURN_IF_ERROR(cudaMemsetAsync(mask_data, 0, mask->SizeInBytes(), Stream()));
+        }
+      }
+    } else if (DataTypeImpl::GetType<TensorSeq>() == X_ml_type) {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/tensor/identity_op.h:48]: >DataTypeImpl::GetType<TensorSeq>() [](http://example.com/codeflow?start=15&length=34)

IsTensorSequence()

File: onnxruntime/core/providers/cuda/tensor/identity_op.h [hunk 3]
```diff
@@ -15,34 +16,57 @@ class IdentityOp final : public CudaKernel {
...
-    const void* source = X->DataRaw(X_type);
-    void* target = Y->MutableDataRaw(X_type);
-    //If source and target pointers are not equal, we need to copy the data.
-    if (target != source) {
-      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(target, source, X->Shape().Size() * X->DataType()->Size(), cudaMemcpyDeviceToDevice, Stream()));
-    }
+      const void* source = X->DataRaw(X_type);
+      void* target = Y->MutableDataRaw(X_type);
+      //If source and target pointers are not equal, we need to copy the data.
+      if (target != source) {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/tensor/identity_op.h:30]: >  if (target != source) { [](http://example.com/codeflow?start=4&length=25)

The kernel has aliasing set, therefore, you will need to check if input and output are the same for sequence as well.

File: onnxruntime/core/providers/cuda/tensor/identity_op.h [hunk 4]
```diff
@@ -15,34 +16,57 @@ class IdentityOp final : public CudaKernel {
...
+      auto X_type = X->DataType();
+      Y->SetType(X_type);
+      AllocatorPtr alloc;
+      auto status = context->GetTempSpaceAllocator(&alloc);
+      if (!status.IsOK()) {
+        ORT_THROW("Unable to get an allocator");
       }
+      std::vector<Tensor> tensors;
+      for (auto iter = X->begin(); iter != X->end(); ++iter) {
+        Tensor tensor(X_type, onnxruntime::TensorShape(iter->Shape()), alloc);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/tensor/identity_op.h:61]: >alloc [](http://example.com/codeflow?start=71&length=5)

I do not think you can use temp allocator to actually allocate output. There is a shortcoming in TensorSequence design

File: onnxruntime/core/providers/cuda/tensor/identity_op.h (hunk 1 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/tensor/identity_op.h:20]: Thanks - add corresponding interface in provider interface.

File: onnxruntime/core/providers/cuda/tensor/identity_op.h (hunk 4 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/tensor/identity_op.h]: That's right, resort to reference of tensor return from X.

File: onnxruntime/core/providers/cuda/tensor/identity_op.h (hunk 3 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/tensor/identity_op.h:29]: Every tensor we copy to are created on the fly, there wouldn't be chance that source equals to target.

File: onnxruntime/core/providers/cuda/tensor/identity_op.h (hunk 1 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/tensor/identity_op.h]: Good point, thanks!

File: onnxruntime/core/providers/cuda/tensor/identity_op.h (hunk 2 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/tensor/identity_op.h]: Done!

File: onnxruntime/core/providers/shared_library/provider_interfaces.h [hunk 5]
```diff
@@ -1580,7 +1604,6 @@ struct Tensor final {
   bool IsDataType() const;
 
   Tensor() = delete;
-  Tensor(const Tensor&) = delete;
```
Comment (RyanUnderhill) [inline onnxruntime/core/providers/shared_library/provider_interfaces.h:1583]: Why was this removed? You can't copy construct tensors by value

File: onnxruntime/core/providers/shared_library/provider_interfaces.h [hunk 6]
```diff
@@ -1663,6 +1686,17 @@ inline const BFloat16* Tensor::Data<BFloat16>() const { return g_host->Tensor__D
 template <>
 inline const MLFloat16* Tensor::Data<MLFloat16>() const { return g_host->Tensor__Data_MLFloat16(this); }
 
+//TensorSee
```
Comment (RyanUnderhill) [inline onnxruntime/core/providers/shared_library/provider_interfaces.h:1689]: Typo, TensorSee->TensorSeq

File: onnxruntime/core/framework/provider_bridge_ort.cc [hunk 7]
```diff
@@ -749,6 +753,14 @@ struct ProviderHostImpl : ProviderHost {
   int32_t Tensor__GetElementType(const Tensor* p) override { return p->GetElementType(); }
   MLDataType Tensor__DataType(const Tensor* p) override { return p->DataType(); }
 
+  // TensorSeq
```
Comment (RyanUnderhill) [inline onnxruntime/core/framework/provider_bridge_ort.cc:756]: // TensorSeq (wrapped)
(since you're wrapping the type vs directly resolving externals)

File: onnxruntime/core/providers/shared_library/provider_interfaces.h (hunk 5 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/shared_library/provider_interfaces.h]: Recovered, thanks!

File: onnxruntime/core/providers/shared_library/provider_interfaces.h (hunk 6 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/shared_library/provider_interfaces.h]: Done.

File: onnxruntime/core/framework/provider_bridge_ort.cc (hunk 7 above)
Comment (RandySheriffH) [reply onnxruntime/core/framework/provider_bridge_ort.cc]: Done.

**Review by RyanUnderhill — APPROVED** [review]
